goth/runner/__init__.py

```diff
@@ -217,31 +221,31 @@ async def _start_nodes(self):
 
         # Start the probes' containers and obtain their IP addresses
         for probe in self.probes:
-            await probe.start()
-            assert probe.ip_address
-            ip = probe.ip_address
-            port = probe.container.ports
-            node_names[ip] = probe.name
-            ports[ip] = port
+            ip_address = await self._exit_stack.enter_async_context(probe.run())
+            node_names[ip_address] = probe.name
+            container_ports = probe.container.ports
+            ports[ip_address] = container_ports
             logger.debug(
                 "Probe for %s started on IP: %s with port mapping: %s",
                 probe.name,
-                ip,
-                port,
+                ip_address,
+                container_ports,
             )
 
         node_names[self.host_address] = "Pytest-Requestor-Agent"
 
+        # Stopping the proxy triggers evaluation of assertions at "the end of events".
+        # Install a callback to to check for assertion failures after the proxy stops.
+        self._exit_stack.callback(self.check_assertion_errors)
+
         # Start the proxy node. The containers should not make API calls
         # up to this point.
         self.proxy = Proxy(
             node_names=node_names,
             ports=ports,
             assertions_module=self.api_assertions_module,
         )
-        self.proxy.start()
-        # Wait for proxy to start. TODO: wait for a log line?
-        await asyncio.sleep(2.0)
```

Comment:
`Proxy.start()` now waits until the `mitmproxy` thread starts its internal server, this should be enough to make sure that the proxy is ready to handle requests.